[guilib] Fixed: don't unfocus control if it's the one being focused

garbear opened the pull request:
I ran across this working on the controller dialog (#8807). When you're mapping a feature, the dialog focuses the feature's button control and asks for input. I wanted the mouse to quickly interrupt the mapping process, so when it gives focus to another control, an "unfocus" command is sent to the feature's button control, canceling the input prompt.

However, if an already-focused control is told to focus, it will first be sent an "unfocus" message, followed by a "focus" message. The "unfocus" message prematurely aborts the input prompt. If a dialog begins prompting for a feature that is already focused, the prompt will be immediately terminated by the "unfocus" message.

This change removes the "unfocus" message when an already-focused control is told to focus, allowing the user to map buttons in the controller dialog.

xhaggi commented:
makes sense but does this introduce any issues for skinners. @ronie mind taking a look.

ronie commented:
perhaps @HitcherUK / @phil65 could chime in.
both of them have been playing with the controller dialog iirc.

phil65 commented:
Just to understand this correctly: Is the only effect that the ```<onunfocus>``` actions do not trigger anymore when control is already focused or does this have some wider effect?

In general this sounds good to me, didnt runtime-test yet though.

garbear commented:
thanks @akva2, didn't think of that use case. The unfocus is necessary there, so this PR isn't the right way to go. It'll add some complexity to the controller dialog, but i'd rather not break existing code.

garbear commented:
what if we added a "refocus" message?

xhaggi commented:
the question is why do we need to send the unfocus/focus msg if the current control never lose its focus?

garbear commented:
> the question is why do we need to send the unfocus/focus msg if the current control never lose its focus?

The answer: buggy behavior. Somewhere, deep in the code, untouched since the Xbox days, is some logic that loses its shit if it sees two focus msgs in a row.

The next question is do we want to make this change and flush out this behavior? I'm perfectly ok exposing a bug if it keeps my controller dialog simpler. But if we want to take the safe road and abandon the change, I'll figure out how to make the controller dialog work around this.

Razzeee commented:
We're pre alpha, so you can't probably find a better timing for flushing it out.

xhaggi commented:
@garbear we should do it the right way and do not introduce another workaround.

garbear commented:
The right way is to not send an unfocus message when a focused control is being focused. This smells like a workaround itself. And I'm against adding a workaround for a workaround.

I've gone through the code and it looks like all unfocus messages boil down to [CGUIControl::SetFocus()](https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/GUIControl.cpp#L328) called for various controls. You can see that @akva2 is right, it does retrigger animations.

I think the correct solution is to add a "refocus" message that explicitly retriggers animations, because otherwise there's no way to indicate that a control is being refocused.

garbear committed:
[guilib] Fixed: don't unfocus control if it's the one being focused

This change adds a check to CGUIWindow to compare the controls receiving
messages. If the controls are the same, only the "focus" message is sent and
the "unfocus" message is dropped.

garbear commented:
PR has been updated. The commit description does a good job of explaining things:

```
This change adds a check to CGUIWindow to compare the controls receiving
messages. If the controls are the same, only the "focus" message is sent and
the "unfocus" message is dropped.

Currently, CGUIControl has a quirk where it requires two messages to
retrigger animations. It must be unfocused first, and then focused again.

Effectively, this quirk has been moved out of CGUIWindow and into
CGUIControl. This is a more appropriate location and prevents the quirk from
bleeding into the messaging system.
```

garbear commented:
OK, I tried a couple approaches, and was always defeated because the messaging is so intertwined with the graphics. I was able to work around this in the controller dialog without much added complexity, so I'll leave this untouched for now.
